aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/SynchronousNonBlockingStepTest.java

```diff
@@ -0,0 +1,245 @@
+
+        // Wait for syncnonblocking to be started
+        System.out.println("Waiting to syncnonblocking to start...");
+        // waitinterrupt is a special key that makes syncnonblocking step to wait inside a while  loop checking
+        // for Thread.interrupted()
+        SynchronousNonBlockingStep.waitForStart("waitinterrupt", b);
+
+        // At this point syncnonblocking is looping and checking for interruption
+
+        CpsFlowExecution e = (CpsFlowExecution) b.getExecutionPromise().get();
+        // Let's force a call to stop. This will try to send an interruption to the run Thread
+        e.interrupt(Result.ABORTED);
+        System.out.println("Looking for interruption received log message");
+        j.waitForMessage("Run thread interrupted", b);
+        j.waitForCompletion(b);
+    }
+
+    @Test
+    public void parallelTest() throws Exception {
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("node {\n" +
+            "echo 'First message'\n" +
+            "parallel( a: { syncnonblocking 'wait0'; echo 'a branch'; }, b: { echo 'b branch' } )\n" +
+            "echo 'Second message'\n" +
+        "}"));
+        WorkflowRun b = p.scheduleBuild2(0).getStartCondition().get();
+
+        SynchronousNonBlockingStep.waitForStart("wait0", b);
+
+        // Wait for "b" branch to print its message
+        j.waitForMessage("b branch", b);
+        System.out.println("b branch finishes");
+
+        // Check that "a" branch is effectively blocked
+        j.assertLogNotContains("a branch", b);
+
+        // Notify "a" branch
+        System.out.println("Continue on wait0");
+        SynchronousNonBlockingStep.notify("wait0");
+
+        // Wait for "a" branch to finish
+        j.waitForMessage("a branch", b);
+        j.waitForCompletion(b);
+    }
+
+    public static final class SynchronousNonBlockingStep extends AbstractStepImpl implements Serializable {
+
+        public static final class State {
+            private static State state;
+            static synchronized State get() {
+                if (state == null) {
+                    state = new State();
+                    return state;
+                }
+                return state;
+            }
+            private State() {}
+            final Set<String> started = new HashSet<String>();
+        }
+
+        private String id;
+
+        @DataBoundConstructor
+        public SynchronousNonBlockingStep(String id) {
+            this.id = id;
+        }
+
+        public String getId() {
+            return id;
+        }
+
+        public static void waitForStart(String id, Run<?,?> b) throws IOException, InterruptedException {
+            State s = State.get();
+            synchronized (s) {
+                while (!s.started.contains(id)) {
+                    if (b != null && !b.isBuilding()) {
+                        throw new AssertionError(JenkinsRule.getLog(b));
```

Comment:
This will throw an error if the build [has not started yet](https://github.com/jenkinsci/workflow-plugin/blob/5c3cbc2b23853cadd94efacb9616761858b8341f/job/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L476), a race condition.